File: backend/app/schemas/instruction.py
"""Data models for instruction stream generation.

Defines data structures such as the instruction format, generation requests and responses.
"""

from typing import Any, Dict, List, Optional, Literal
from pydantic import BaseModel, Field


# ==================== Instruction Format Definitions ====================

class InstructionBase(BaseModel):
    """Instruction base class"""
    op: str = Field(..., description="Instruction operation type")


class SetInstruction(InstructionBase):
    """Set field value instruction"""
    op: Literal["set"] = "set"
    path: str = Field(..., description="Field path in JSON Pointer format, e.g. /name or /config/theme")
    value: Any = Field(..., description="Value to set, can be any type")


class AppendInstruction(InstructionBase):
    """Append element to array instruction"""
    op: Literal["append"] = "append"
    path: str = Field(..., description="Array path in JSON Pointer format")
    value: Any = Field(..., description="Element to append")


class DoneInstruction(InstructionBase):
    """Generation completion flag instruction"""
    op: Literal["done"] = "done"


# Union type: all instruction types
Instruction = SetInstruction | AppendInstruction | DoneInstruction


# ==================== Generation Config ====================

class GenerationConfig(BaseModel):
    """Card generation config

    Defines the strategy and prompts for generating card content.
    """
    mode: Literal["instruction_stream"] = Field(
        default="instruction_stream",
        description="Generation mode, currently only instruction stream mode is supported"
    )
    prompt_template: Optional[str] = Field(
        default=None,
        description="Global prompt template (optional)"
    )
    field_hints: Optional[Dict[str, str]] = Field(
        default=None,
        description="Field-level generation hints, key is field path, value is hint text"
    )
    field_order: Optional[List[str]] = Field(
        default=None,
        description="Suggested field generation order"
    )
    custom: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Custom config (for extension)"
    )


# ==================== API Request/Response Models ====================

class ConversationMessage(BaseModel):
    """Conversation message"""
    role: Literal["system", "user", "assistant"] = Field(..., description="Message role")
    content: str = Field(..., description="Message content")


class InstructionGenerateRequest(BaseModel):
    """Instruction stream generation request"""

    # LLM config
    llm_config_id: int = Field(..., description="LLM config ID")

    # User input
    user_prompt: str = Field(default="", description="User-provided prompt or reply")

    # Schema definition
    response_model_schema: Dict[str, Any] = Field(..., description="JSON Schema of the target data structure")

    # Current data state
    current_data: Dict[str, Any] = Field(default_factory=dict, description="Currently generated data")

    # Conversation context
    conversation_context: List[ConversationMessage] = Field(
        default_factory=list,
        description="Conversation history (maintained by frontend)"
    )

    # Generation config (optional)
    generation_config: Optional[GenerationConfig] = Field(
        default=None,
        description="Generation config, uses default config if empty"
    )

    # Prompt template (optional, overrides default)
    prompt_template: Optional[str] = Field(
        default=None,
        description="Custom prompt template"
    )

    # Context info (optional)
    context_info: Optional[str] = Field(
        default=None,
        description="Context injection info (e.g. related entities, existing cards)"
    )

    # Sampling parameters
    temperature: Optional[float] = Field(default=0.7, description="Sampling temperature")
    max_tokens: Optional[int] = Field(default=None, description="Max generation token count")
    timeout: Optional[float] = Field(default=150, description="Timeout (seconds)")

    # Dependency data (e.g. entity name list)
    deps: Optional[str] = Field(default=None, description="Dependency data, used for validation")


# ==================== SSE Event Types ====================

class ThinkingEvent(BaseModel):
    """Thinking event (AI's natural language output)"""
    type: Literal["thinking"] = "thinking"
    text: str = Field(..., description="Thinking content or question")


class InstructionEvent(BaseModel):
    """Instruction event (validated instruction)"""
    type: Literal["instruction"] = "instruction"
    instruction: Instruction = Field(..., description="Instruction object")


class WarningEvent(BaseModel):
    """Warning event (non-fatal error)"""
    type: Literal["warning"] = "warning"
    text: str = Field(..., description="Warning message")


class ErrorEvent(BaseModel):
    """Error event (fatal error)"""
    type: Literal["error"] = "error"
    text: str = Field(..., description="Error message")


class DoneEvent(BaseModel):
    """Done event"""
    type: Literal["done"] = "done"
    success: bool = Field(default=True, description="Whether completed successfully")
    message: Optional[str] = Field(default=None, description="Completion message")


# Union type: all event types
StreamEvent = ThinkingEvent | InstructionEvent | WarningEvent | ErrorEvent | DoneEvent

File: frontend/src/renderer/src/types/instruction.ts
/**
 * Type definitions related to instruction stream generation
 * 
 * Kept in sync with the backend Pydantic models
 */

// ==================== Instruction format definitions ====================

/**
 * Instruction operation type
 */
export type InstructionOp = 'set' | 'append' | 'done'

/**
 * Base instruction interface
 */
export interface InstructionBase {
  op: InstructionOp
}

/**
 * Set field value instruction
 */
export interface SetInstruction extends InstructionBase {
  op: 'set'
  path: string  // JSON Pointer format, e.g. /name or /config/theme
  value: any    // The value to set
}

/**
 * Append element to array instruction
 */
export interface AppendInstruction extends InstructionBase {
  op: 'append'
  path: string  // JSON Pointer format array path
  value: any    // The element to append
}

/**
 * Generate completion flag instruction
 */
export interface DoneInstruction extends InstructionBase {
  op: 'done'
}

/**
 * Instruction union type
 */
export type Instruction = SetInstruction | AppendInstruction | DoneInstruction

// ==================== Generation config ====================

/**
 * Card generation config
 */
export interface GenerationConfig {
  mode?: 'instruction_stream'
  prompt_template?: string
  field_hints?: Record<string, string>
  field_order?: string[]
  custom?: Record<string, any>
}

// ==================== API request/response models ====================

/**
 * Conversation message
 */
export interface ConversationMessage {
  role: 'system' | 'user' | 'assistant'
  content: string
}

/**
 * Instruction stream generation request
 */
export interface InstructionGenerateRequest {
  llm_config_id: number
  user_prompt?: string
  response_model_schema: Record<string, any>
  current_data?: Record<string, any>
  conversation_context?: ConversationMessage[]
  generation_config?: GenerationConfig
  prompt_template?: string
  context_info?: string
  temperature?: number
  max_tokens?: number
  timeout?: number
  deps?: string
}

// ==================== SSE event types ====================

/**
 * Thinking event (AI's natural language output)
 */
export interface ThinkingEvent {
  type: 'thinking'
  text: string
}

/**
 * Instruction event (validated instruction)
 */
export interface InstructionEvent {
  type: 'instruction'
  instruction: Instruction
}

/**
 * Warning event (non-fatal error)
 */
export interface WarningEvent {
  type: 'warning'
  text: string
}

/**
 * Error event (fatal error)
 */
export interface ErrorEvent {
  type: 'error'
  text: string
}

/**
 * Done event
 */
export interface DoneEvent {
  type: 'done'
  success?: boolean
  message?: string
}

/**
 * Stream event union type
 */
export type StreamEvent = ThinkingEvent | InstructionEvent | WarningEvent | ErrorEvent | DoneEvent

// ==================== Generation panel message types ====================

/**
 * Generation panel message type
 */
export type GenerationMessageType = 'thinking' | 'action' | 'system' | 'user' | 'warning' | 'error'

/**
 * Generation panel message
 */
export interface GenerationMessage {
  type: GenerationMessageType
  content: string
  timestamp: number
}
